src/transformers/trainer.py

```diff
@@ -1588,17 +1604,30 @@ def _inner_training_loop(
                 # May be slightly incorrect if the last batch in the training dataloader has a smaller size but it's
                 # the best we can do.
                 num_train_samples = args.max_steps * total_train_batch_size
+                num_train_tokens = (
+                    args.max_steps
+                    * self.num_tokens(train_dataloader, True)
+                    * args.gradient_accumulation_steps
+                    * args.world_size
+                )
             else:
                 max_steps = math.ceil(args.num_train_epochs * num_update_steps_per_epoch)
                 num_train_epochs = math.ceil(args.num_train_epochs)
                 num_train_samples = self.num_examples(train_dataloader) * args.num_train_epochs
+                num_train_tokens = self.num_tokens(train_dataloader) * args.world_size * args.num_train_epochs
```

Comment:
We are calculating `max_steps` on line 1614. Why can't we use that to avoid iterating through the entire dataloader?

Reply:
`max_steps` doesnot contains the num of `tokens` infomation, so as call `num_tokens`.  

yes, the iteration will slow down the whole training ,especally big datasets. Currently i am trying moving the approximation of tokens per second into each training step so as to avoild duplicated dataloading process. maybe that will solve the common concerns.

Reply:
If that's not possible (which I don't think it is without some effort) I'd rather see this as an opt-in. Because the other issue with this is users can use the `Trainer` to train any dataset, not just text (though that's the most common). So again, I'd rather see this as an opt-in specifying more.

Reply:
yeah.i got u:-) an opt-in maybe the most direct way for user to choose open `tgs` or not.